### PR TITLE
SharedMemory Status server API support

### PR DESCRIPTION
**Pull request by CoderHam**
- [x] fix shared memory unregisterAll server implementation

### CODE REVIEW

File: src/core/grpc_service.proto [hunk 1]
```diff
@@ -325,6 +325,12 @@ message SharedMemoryControlRequest
     //@@       To unregister all active shared memory regions.
     //@@
     UNREGISTER_ALL = 2;
+
```
Comment (deadeyegoodwin) [inline src/core/grpc_service.proto]: Having one request type for all the control requests doesn't work if the different requests need different inputs. Instead look at the TraceControl messages to see how you can use oneof to define different request (and response) for different request types.

File: src/core/server.cc [hunk 2]
```diff
@@ -415,6 +416,12 @@ InferenceServer::SharedMemoryAddress(
       name, offset, byte_size, shm_mapped_addr);
 }
 
+Status
+InferenceServer::GetSharedMemoryStatus(SharedMemoryControlResponse* shm_status)
+{
+  return shared_memory_manager_->GetSharedMemoryStatus(shm_status);
```
Comment (deadeyegoodwin) [inline src/core/server.cc]: Need to check ready_state_ and update inflight_request_counter_ as in UnregisterAllSharedMemory.

File: src/core/server.h [hunk 3]
```diff
@@ -38,6 +38,7 @@
 #include "src/core/provider.h"
 #include "src/core/server_status.h"
 #include "src/core/server_status.pb.h"
+#include "src/core/grpc_service.pb.h"
```
Comment (deadeyegoodwin) [inline src/core/server.h]: You need to run the formatter

File: src/core/shared_memory_manager.cc [hunk 4]
```diff
@@ -205,6 +205,17 @@ SharedMemoryManager::UnregisterAllSharedMemory()
   return Status::Success;
 }
 
+Status
+SharedMemoryManager::GetSharedMemoryStatus(SharedMemoryControlResponse* shm_status)
```
Comment (deadeyegoodwin) [inline src/core/shared_memory_manager.cc]: This needs an implementation.

File: src/core/trtserver.cc [hunk 5]
```diff
@@ -1154,6 +1156,36 @@ TRTSERVER_ServerSharedMemoryAddress(
   return nullptr;  // success
 }
 
+TRTSERVER_Error*
+TRTSERVER_ServerGetSharedMemoryStatus(
```
Comment (deadeyegoodwin) [inline src/core/trtserver.cc]: ServerSharedMemoryStatus

File: src/core/grpc_service.proto (hunk 1 above)
Comment (CoderHam) [reply src/core/grpc_service.proto]: Done

File: src/clients/c++/request_grpc.cc [hunk 6]
```diff
@@ -333,58 +341,79 @@ SharedMemoryControlGrpcContextImpl::RegisterSharedMemory(
...
 
 Error
 SharedMemoryControlGrpcContextImpl::UnregisterAllSharedMemory()
 {
-  return SendRequest("", SharedMemoryControlRequest::UNREGISTER_ALL, "", 0, 0);
+  return SendRequest("", UNREGISTER_ALL, "", 0, 0, nullptr);
 }
 
 Error
 SharedMemoryControlGrpcContextImpl::SendRequest(
```
Comment (deadeyegoodwin) [inline src/clients/c++/request_grpc.cc]: Why use a common routine? There is not that much common code across the request types and it just makes it confusing.

File: src/core/grpc_service.proto [hunk 7]
```diff
@@ -370,47 +370,69 @@ message InferSharedMemoryRegion
...
 {
-  //@@  .. cpp:enum:: Type
+  //@@  .. cpp:var:: message RegisterInfo
   //@@
-  //@@     Types of control operations for shared memory
+  //@@     The information needed to register the shared memory region.
   //@@
-  enum Type {
-    //@@    .. cpp:enumerator:: Type::REGISTER = 0
+  message RegisterInfo
```
Comment (deadeyegoodwin) [inline src/core/grpc_service.proto]: Drop "info" here and elsewhere. Just "Register".

File: src/core/grpc_service.proto [hunk 8]
```diff
@@ -370,47 +370,69 @@ message InferSharedMemoryRegion
 //@@
 message SharedMemoryControlRequest
 {
-  //@@  .. cpp:enum:: Type
+  //@@  .. cpp:var:: message RegisterInfo
   //@@
-  //@@     Types of control operations for shared memory
+  //@@     The information needed to register the shared memory region.
```
Comment (deadeyegoodwin) [inline src/core/grpc_service.proto]: "Register a shared memory region".

File: src/core/grpc_service.proto [hunk 9]
```diff
@@ -370,47 +370,69 @@ message InferSharedMemoryRegion
...
   //@@
-  //@@  .. cpp:var:: Type type
+  //@@     Information needed to UnregisterAll and GetStatus of shared memory.
   //@@
-  //@@     The control type that states whether to register/unregister the
-  //@@     specified shared memory region. Unregister all active shared memory
-  //@@     regions or get the list of active shared memory regions
-  //@@
-  Type type = 1;
+  message NoInfo {}
```
Comment (deadeyegoodwin) [inline src/core/grpc_service.proto]: Make an "UnregisterAll" and "Status" message. Both will be empty.

File: src/core/grpc_service.proto [hunk 10]
```diff
@@ -370,47 +370,69 @@ message InferSharedMemoryRegion
...
+
+    //@@    .. cpp:var:: NoInfo unregister_all
+    //@@
+    //@@       To unregister all active shared memory regions.
+    //@@
+    NoInfo unregister_all = 3;
+
+    //@@    .. cpp:var:: NoInfo get_status
+    //@@
+    //@@       To get the list of all active shared memory regions.
```
Comment (deadeyegoodwin) [inline src/core/grpc_service.proto]: Not a "list". "Get the status of all active shared memory regions".

File: src/core/grpc_service.proto [hunk 11]
```diff
@@ -426,6 +448,13 @@ message SharedMemoryControlResponse
   //@@     The status of the request, indicating success or failure.
   //@@
   RequestStatus request_status = 1;
+
+  //@@
+  //@@  .. cpp:var:: SharedMemoryRegion shared_memory_region
+  //@@
+  //@@     The list of active/registered shared memory regions.
+  //@@
+  repeated InferSharedMemoryRegion shared_memory_region = 2;
```
Comment (deadeyegoodwin) [inline src/core/grpc_service.proto]: This is only relevant for "Status" so again so a oneof with a "Status" message where in that message is this repeated. Also status results should be returned as a SharedMemoryStatus message, right?

File: src/servers/grpc_server.cc [hunk 12]
```diff
@@ -1476,40 +1476,56 @@ SharedMemoryControlHandler::Process(Handler::State* state, bool rpc_ok)
...
+      err = TRTSERVER_ServerSharedMemoryStatus(
+          trtserver_.get(), &shm_status_protobuf);
+      if (err == nullptr) {
+        const char* status_buffer;
+        size_t status_byte_size;
+        err = TRTSERVER_ProtobufSerialize(
+            shm_status_protobuf, &status_buffer, &status_byte_size);
         if (err == nullptr) {
-          err = TRTSERVER_ServerUnregisterAllSharedMemory(trtserver_.get());
+          if (!response.ParseFromArray(status_buffer, status_byte_size)) {
```
Comment (deadeyegoodwin) [inline src/servers/grpc_server.cc]: How can this be correct. The response is not just the SharedMemoryStatus message.

File: src/core/grpc_service.proto [hunk 13]
```diff
@@ -370,47 +370,69 @@ message InferSharedMemoryRegion
...
-    //@@    .. cpp:enumerator:: Type::REGISTER = 0
+  message RegisterInfo
+  {
+    //@@  .. cpp:var:: InferSharedMemoryRegion shared_memory_region
     //@@
-    //@@       To register the specified shared memory region.
+    //@@     The shared memory region to register.
     //@@
-    REGISTER = 0;
+    InferSharedMemoryRegion shared_memory_region = 5;
```
Comment (deadeyegoodwin) [inline src/core/grpc_service.proto]: InferSharedMemoryRegion is no longer needed as a separate mesage. The contents of that message should become the contents of Register.

File: src/core/shared_memory_manager.cc [hunk 14]
```diff
@@ -149,11 +149,8 @@ SharedMemoryManager::RegisterSharedMemory(
 }
 
 Status
-SharedMemoryManager::UnregisterSharedMemory(const std::string& name)
+SharedMemoryManager::UnregisterSharedMemoryHelper(const std::string& name)
 {
```
Comment (deadeyegoodwin) [inline src/core/shared_memory_manager.cc]: Add a comment that the lock must be held when this function is called.

File: src/core/shared_memory_manager.cc [hunk 15]
```diff
@@ -205,6 +211,23 @@ SharedMemoryManager::UnregisterAllSharedMemory()
   return Status::Success;
```
Comment (deadeyegoodwin) [inline src/core/shared_memory_manager.cc]: Above you shouldn't just log an error you should also return the same message as an error status.

File: src/core/shared_memory_manager.h [hunk 16]
```diff
@@ -83,8 +83,15 @@ class SharedMemoryManager {
       const std::string& name, const std::string& shm_key, const size_t offset,
       const size_t byte_size);
 
+  /// Helper function that unregisters a specified shared memory region if
+  /// registered else do nothing and return success.
+  /// \param name The user-given name for the shared memory region to be
+  /// registered.
+  /// \return error status.
+  Status UnregisterSharedMemoryHelper(const std::string& name);
```
Comment (deadeyegoodwin) [inline src/core/shared_memory_manager.h]: This should be private.

File: src/servers/grpc_server.cc [hunk 17]
```diff
@@ -1476,40 +1476,56 @@ SharedMemoryControlHandler::Process(Handler::State* state, bool rpc_ok)
...
-        err = smb_manager_->Remove(&smb, request.shared_memory_region().name());
-        if ((err == nullptr) && (smb != nullptr)) {
-          err = TRTSERVER_ServerUnregisterSharedMemory(trtserver_.get(), smb);
-          TRTSERVER_Error* del_err = TRTSERVER_SharedMemoryBlockDelete(smb);
-          if (del_err != nullptr) {
-            LOG_ERROR << "failed to delete shared memory block: "
-                      << TRTSERVER_ErrorMessage(del_err);
-            TRTSERVER_ErrorDelete(del_err);
-          }
+    if (request.has_register_()) {
```
Comment (deadeyegoodwin) [inline src/servers/grpc_server.cc]: has_register_? Does that function even exist?

File: src/servers/grpc_server.cc [hunk 18]
```diff
@@ -1476,40 +1476,56 @@ SharedMemoryControlHandler::Process(Handler::State* state, bool rpc_ok)
...
+          }
         }
-        break;
-      default:
-        err = TRTSERVER_ErrorNew(
-            TRTSERVER_ERROR_UNKNOWN,
-            "unknown sharedmemorycontrol request type");
-        break;
+      }
+      if (err != nullptr) {
```
Comment (deadeyegoodwin) [inline src/servers/grpc_server.cc]: Why log an error here? We don't log for other request failures. We just return the error status.

File: src/servers/http_server.cc [hunk 19]
```diff
@@ -687,8 +700,49 @@ HTTPAPIServer::HandleSharedMemoryControl(
...
+                "failed to parse shared memory status");
+          } else {
+            std::string shm_status_str = shm_status.DebugString();
+            evbuffer_add(
+                req->buffer_out, shm_status_str.c_str(), shm_status_str.size());
+          }
+        }
+      }
+    }
+    if (err != nullptr) {
```
Comment (deadeyegoodwin) [inline src/servers/http_server.cc]: Remove this logging

File: src/core/grpc_service.proto (hunk 7 above)
Comment (CoderHam) [reply src/core/grpc_service.proto]: Sure

File: src/clients/c++/request_grpc.cc (hunk 6 above)
Comment (CoderHam) [reply src/clients/c++/request_grpc.cc]: Will split this into different components and remove the SendRequest function

File: src/servers/grpc_server.cc (hunk 17 above)
Comment (CoderHam) [reply src/servers/grpc_server.cc]: register gets named as register_ due to a keyword issue or conflict.

File: src/core/grpc_service.proto (hunk 11 above)
Comment (CoderHam) [reply src/core/grpc_service.proto]: That is correct

File: src/servers/grpc_server.cc (hunk 12 above)
Comment (CoderHam) [reply src/servers/grpc_server.cc]: I add the request status later. Not entirely sure why the ParseFromArray does not throw an error though.

File: src/core/grpc_service.proto (hunk 13 above)
Comment (CoderHam) [reply src/core/grpc_service.proto]: Deleted InferSharedMemoryRegion. Reuses SharedMemoryRegion from server_status.proto

File: src/core/grpc_service.proto [hunk 20]
```diff
@@ -370,47 +370,75 @@ message InferSharedMemoryRegion
...
-  //@@     specified shared memory region. Unregister all active shared memory
-  //@@     regions or get the list of active shared memory regions
+  message UnregisterAll {}
+
+  //@@  .. cpp:var:: message GetStatus
+  //@@
+  //@@     Get the status of all active shared memory regions.
   //@@
-  Type type = 1;
+  message GetStatus {}
```
Comment (deadeyegoodwin) [inline src/core/grpc_service.proto]: Just "Status"

File: src/core/grpc_service.proto (hunk 13 above)
Comment (deadeyegoodwin) [reply src/core/grpc_service.proto]: This comment seems wrong. I still see InferSharedMemoryRegion. It should be deleted and contents moved to Register. SharedMemoryRegion is the status and in the future we may add more things to the status so don't reuse that here.

File: src/core/grpc_service.proto (hunk 11 above)
Comment (deadeyegoodwin) [reply src/core/grpc_service.proto]: So this needs to be a oneof which (currently) just has a "Status" message which includes SharedMemoryStatus.

**Comment by CoderHam** [comment]
Register is re-used in the SharedMemoryStatus. Also the SharedMemoryStatus Response should be Status + repeated SharedMemoryRegions

File: src/core/grpc_service.proto [hunk 21]
```diff
@@ -314,12 +328,37 @@ message SharedMemoryControlRequest
 //@@
 message SharedMemoryControlResponse
 {
+  //@@
+  //@@.. cpp:var:: message SharedMemoryStatusResponse
+  //@@
+  //@@   Status of all active shared memory regions.
+  //@@
+  message SharedMemoryStatusResponse
```
Comment (deadeyegoodwin) [inline src/core/grpc_service.proto]: Just "Status" to parallel what is done on the request side.

File: src/core/grpc_service.proto [hunk 22]
```diff
@@ -314,12 +328,37 @@ message SharedMemoryControlRequest
...
+  }
+
   //@@
   //@@  .. cpp:var:: RequestStatus request_status
   //@@
   //@@     The status of the request, indicating success or failure.
   //@@
   RequestStatus request_status = 1;
+
+  oneof status_response
```
Comment (deadeyegoodwin) [inline src/core/grpc_service.proto]: oneof shared_memory_control

**Pull request merged**